lib/core/refresh.js

```diff
@@ -0,0 +1,246 @@
+  }
+}
+
+const TokenExpirationStatusEnum = Object.freeze({
+  UNKNOWN: 'UNKNOWN',
+  VALID: 'VALID',
+  EXPIRED: 'EXPIRED',
+  REFRESH_EXPIRED: 'REFRESH_EXPIRED'
+})
+
+export class TokenExpirationStatus {
+  constructor (scheme) {
+    this.$scheme = scheme
+  }
+
+  _calculateTokenStatus () {
+    const now = Date.now()
+    let tokenExpiresAt, refreshTokenExpiresAt
+
+    try {
+      tokenExpiresAt = this.$scheme._getTokenExpiration()
+      refreshTokenExpiresAt = this.$scheme._getRefreshTokenExpiration()
+
+      if (!tokenExpiresAt) {
+        return TokenExpirationStatusEnum.UNKNOWN
+      }
+    } catch (error) {
+      return TokenExpirationStatusEnum.UNKNOWN
+    }
+
+    // Give us some slack to help the token from expiring between validation and usage
+    const timeSlackMillis = 500
+    tokenExpiresAt -= timeSlackMillis
+    refreshTokenExpiresAt -= timeSlackMillis
+
+    // Token is still valid
+    if (now < tokenExpiresAt) {
+      return TokenExpirationStatusEnum.VALID
+    }
+
+    if (now > refreshTokenExpiresAt) {
+      return TokenExpirationStatusEnum.REFRESH_EXPIRED
+    }
+
+    return TokenExpirationStatusEnum.EXPIRED
+  }
+
+  setTokenExpiration (result = {}) {
+    const token = this.$scheme.$auth.getToken(this.$scheme.name)
+    let decodedToken, tokenExpiration
+    const _tokenIssuedAt = getProp(result, this.$scheme.options.issuedAt) || Date.now()
+    const _tokenTTL = getProp(result, this.$scheme.options.expiresIn) || this.$scheme.options.token.maxAge
+    const _tokenExpiresAt = getProp(result, this.$scheme.options.expiresAt) * 1000 || _tokenIssuedAt + (_tokenTTL * 1000)
+
+    try {
+      decodedToken = jwtDecode(token)
+      tokenExpiration = decodedToken.exp * 1000
+    } catch (error) {
+      // If the token is not jwt, we can't decode and refresh it, use _tokenExpiresAt value
+      tokenExpiration = _tokenExpiresAt || false
+
+      if (!(error instanceof InvalidTokenError)) {
+        throw error
+      }
+    }
+
+    // Set token expiration
+    this.$scheme._setTokenExpiration(tokenExpiration)
+
+    // Update status
+    this.syncStatus()
+
+    return tokenExpiration
+  }
+
+  setRefreshTokenExpiration () {
+    const refreshToken = this.$scheme.$auth.getRefreshToken(this.$scheme.name)
+    let decodedToken, refreshTokenExpiration
+    const _tokenIssuedAt = Date.now()
+    const _tokenTTL = this.$scheme.options.refreshToken.maxAge
+    const _tokenExpiresAt = _tokenIssuedAt + (_tokenTTL * 1000)
+
+    try {
+      decodedToken = jwtDecode(refreshToken)
+      refreshTokenExpiration = decodedToken.exp * 1000
+    } catch (error) {
+      // If the token is not jwt, we can't decode and refresh it, use _tokenExpiresAt value
+      refreshTokenExpiration = _tokenExpiresAt || false
+
+      if (!(error instanceof InvalidTokenError)) {
+        throw error
+      }
+    }
+
+    // Set token expiration
+    this.$scheme._setRefreshTokenExpiration(refreshTokenExpiration)
```

Comment:
I couldn't find any other references to `_setRefreshTokenExpiration`. Is it supposed to be `setRefreshTokenExpiration`? If not probably we can add an if check to ensure this function exists on the scheme?

Reply:
Yeah I think we should add a check. `_setRefreshTokenExpiration` is a method in scheme that `setUniversal`.

Reply:
`_` prefixed methods/props should be only called by the same class. If we want to call them should have not `_`

Reply:
Ok, I'll fix this. That happened because I was using `_setRefreshTokenExpiration` in`refresh` scheme before we create the Refresh Controller. I forgot to rename.

Reply:
Maybe we could move both of them to core, as the other methods that uses `syncUniversal`, `setUniversal` and `getUniversal`

Reply:
Indeed If it is common enough between schemes, we can move to this controller :)